lib/elastic_apm/central_config.rb

```diff
@@ -0,0 +1,139 @@
+# frozen_string_literal: true
+
+require 'elastic_apm/central_config/cache_control'
+
+module ElasticAPM
+  # @api private
+  class CentralConfig
+    include Logging
+
+    # @api private
+    class ResponseError < InternalError
+      def initialize(response)
+        @response = response
+      end
+
+      attr_reader :response
+    end
+    class ClientError < ResponseError; end
+    class ServerError < ResponseError; end
+
+    DEFAULT_MAX_AGE = 300
+
+    def initialize(config)
+      @config = config
+      @modified_options = {}
+      @service_info = {
+        'service.name': config.service_name,
+        'service.environment': config.environment
+      }.to_json
+    end
+
+    attr_reader :config, :task
+
+    def start
+      return unless config.central_config?
+
+      fetch_and_apply_config
+    end
+
+    def fetch_and_apply_config
+      @task&.cancel
+      @task =
+        Concurrent::Promise
+        .execute(&method(:fetch_config))
+        .on_success(&method(:handle_success))
+        .rescue(&method(:handle_error))
+    end
+
+    def stop
+      @task&.cancel
+    end
+
+    # rubocop:disable Metrics/MethodLength
+    def fetch_config
+      resp = perform_request
+
+      case resp.status
+      when 200..299
+        resp
+      when 300..399
+        resp
```

Comment:
does that mean we'll call `handle_success` for 3xx? in which case, won't that revert config on 304?

Reply:
As far as I understand it, the 304s will have empty bodies and so `handle_success` will handle it.

Reply:
Oh no, you're right. I'll fix.

Reply:
I don't think we should rely on the 304 body being empty or non-JSON in any case.

Reply:
Agreed. I modified `handle_success` to explicitly handle 304s